clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2142,6 +2142,56 @@ namespace {
       }
    }
 
+   void ResetKilledBounds(StmtDeclSetTy &KilledBounds, Stmt *S,
+                          BoundsContextTy &ObservedBounds) {
+     auto I = KilledBounds.find(S);
+     if (I == KilledBounds.end())
+       return;
+
+     // KilledBounds stores a mapping of statements to all variables whose
+     // bounds are killed by each statement. Here are remove the bounds of all
+     // such variables from ObservedBounds whose bounds are killed by the
+     // statement S. After removal, the bounds of these variables would default
+     // to the user declared bounds in DeclaredBounds.
+     for (const VarDecl *V : I->second)
+       ObservedBounds.erase(V);
```

Comment:
I think this should reset `ObservedBounds[V]` to the normalized declared bounds of `V`, rather than erasing `V` from the context. Before calling `Check` on a statement `S`, the `ObservedBounds` context should contain, for each variable `V` that is in scope at `S`:
1. If `V` has widened bounds `(lb, ub + i`) at `S`, `ObservedBounds[V] => bounds(lb, ub + i)`
2. Otherwise, if `V` has declared bounds `(lb, ub)`, `ObservedBounds[V] => bounds(lb, ub)`

For example:
```
void pointer_widening() {
  nt_array_ptr<char> p : count(0) = "a";

  if (p[0]) {
    // This assignment kills the widened bounds (p, p + 1) of p.
    // Before checking this assignment, ObservedBounds should contain p => (p, p + 0).
    // The original value of p in this assignment is p - 1. p - 1 should replace p in the observed bounds of p.
    // After checking this assignment, ObservedBounds should contain p => (p - 1, p -1 + 0).
    p = p + 1;
  }
}
```